### PR TITLE
Refactor progress handling

### LINKED ISSUE #1050: files progress dialog minimized
Sometime when I copy files interactively using io.elementary.files (either local to to smb/ftp) the copy dialog disappers. When I manage to restore it it is just a little white box so I can't acctually see what's happening or if there are any problems. The file icon in the plank show some (dark) progess bar but the dialog box is just empty.

### LINKED ISSUE #1153: Files process does not end and plank progress bar remains after some file operations.
I have searched open and closed issues for duplicates. Maybe related to #1050.

## Describe the bug
After carrying out some file operations, I noticed that the dock icon still had a progressbar emblem.  Upon closing all Files windows the Files process did not end (as seen in system-monitor). Suggests application has been inhibited but not released.

This does not always happen so maybe a race condition of some kind.

### Expected behavior
After completion or cancellation of file operations the application should be released and the plank icon progressbar removed.

## Platform Information
Hera 5.1 daily
 I'm using the latest version from git that I've manually compiled

## Additional context
I suspect this maybe a regression due to this fix: #727

### PR DESCRIPTION
**Pull request by jeremypw**
Fixes #1050 

- [x] Ensure Progress.Info is finalized by disconnecting signals
* Move inhibit/uninhibit code into Progress.Info
* Use properties with implicit accessors more
* Some better function names
* Progress.InfoManager responsible for creating infos
* Merge some functions

### CODE REVIEW

File: libcore/ProgressInfo.vala [hunk 1]
```diff
@@ -17,25 +17,22 @@
...
-    private bool activity_mode;
-    private bool is_started;
-    private bool is_finished;
-    private bool is_paused;
+    public double progress { get; private set; default = 0; }
+    public double current { get; private set; default = -1; }
+    public double total { get; private set; default = -1; }
+    public bool activity_mode { get; private set; default = true; }
+
+    public uint inhibit_cookie { get; set; default = -1; }
```
Comment (tintou) [inline libcore/ProgressInfo.vala]:
```suggestion
    public uint inhibit_cookie { get; set; default = 0; }
```
It's an uint so -1 gives uint.MAX technically. Also, the documentation on inhibit stands that it returns:
> A non-zero cookie that is used to uniquely identify this request.
So the code should be changed accordingly (meaning that >=0 should be != 0)

File: src/View/Widgets/ProgressInfoWidget.vala [hunk 2]
```diff
@@ -83,19 +69,26 @@ public class Marlin.Progress.InfoWidget : Gtk.Box {
...
+
+        info.changed.connect (update_data);
+        info.progress_changed.connect (update_progress);
+        info.progress_finished.connect (() => { destroy (); });
     }
 
-    private void update_data () {
-        string status = this.info.get_status ();
+    public void update_data () {
+        string status = info.status;
```
Comment (tintou) [inline src/View/Widgets/ProgressInfoWidget.vala]: You can probably do that
```suggestion
        unowned string status = info.status;
```

File: src/View/Widgets/ProgressInfoWidget.vala [hunk 3]
```diff
@@ -83,19 +69,26 @@ public class Marlin.Progress.InfoWidget : Gtk.Box {
...
         (this.status as Gtk.Label).set_text (status);
 
-        string details = this.info.get_details ();
+        string details = info.details;
```
Comment (tintou) [inline src/View/Widgets/ProgressInfoWidget.vala]:
```suggestion
        unowned string details = info.details;
```

File: libcore/ProgressInfo.vala (hunk 1 above)
Comment (jeremypw) [reply libcore/ProgressInfo.vala]: Thanks.  I was just porting the C code but you are right, if a valid cookie is always non-zero then it is not necessary to default to -1.

File: src/View/Widgets/ProgressInfoWidget.vala (hunk 3 above)
Comment (jeremypw) [reply src/View/Widgets/ProgressInfoWidget.vala]: I'll try it.

File: src/View/Widgets/ProgressInfoWidget.vala (hunk 2 above)
Comment (jeremypw) [reply src/View/Widgets/ProgressInfoWidget.vala]: I'll try it.

**Comment by jeremypw** [comment]
Being split into smaller PRs

**Comment by CodingSpiderFox** [comment]
I really really hope this gets fixed soon ;)

**Comment by hanaral** [comment]
I understand that the visual refactoring of the sidebar and listview badges is key right now for the release; but could these PR's get updated and reviewed? a lot of people get turned away from Files due to the file operation bugs
_(I'm not trying to imply that there is somehow a single thing causing these problems; can't image the painstaking effort building a F.M. is)_

**Comment by jeremypw** [comment]
Sorry, this has been put on the back burning while some work has been done on porting some file operation code from C to Vala.  I'll try to get back to it before long.

**Comment by CodingSpiderFox** [comment]
So was this already fixed or will a new attempt to fix it be made?

**Comment by jeremypw** [comment]
The main issue is that there is already a lot of code awaiting review so there is little incentive to spend time adding to the queue.

**Pull request closed without merging**